transport/conn.go

```diff
@@ -191,3 +187,98 @@ func (p *conn) handshake() error {
 	p.open = true
 	return nil
 }
+
+type connHandshakerPipe interface {
+	handshake() error
+
+	Pipe
+}
+
+type connHandshakerItem struct {
+	c connHandshakerPipe
+	e error
+}
+type connHandshaker struct {
+	workq  map[connHandshakerPipe]bool
+	doneq  []*connHandshakerItem
+	closed bool
+	cv     *sync.Cond
+	sync.Mutex
+}
+
+// NewConnHandshaker returns a Handshaker that works with
+// Pipes created via NewConnPipe or NewConnPipeIPC.
+func NewConnHandshaker() Handshaker {
+	h := &connHandshaker{
+		workq:  make(map[connHandshakerPipe]bool),
+		closed: false,
+	}
+	h.cv = sync.NewCond(h)
+	return h
+}
+
+func (h *connHandshaker) Wait() (Pipe, error) {
+	h.Lock()
+	defer h.Unlock()
+	for len(h.doneq) == 0 && !h.closed {
+		h.cv.Wait()
+	}
+	if h.closed {
+		return nil, mangos.ErrClosed
+	}
+	item := h.doneq[0]
+	h.doneq = h.doneq[1:]
+	return item.c, item.e
+}
+
+func (h *connHandshaker) Start(p Pipe) error {
+	conn, ok := p.(connHandshakerPipe)
+	if !ok {
+		p.Close()
+		return mangos.ErrBadTran
+	}
+	h.Lock()
+	defer h.Unlock()
+	if h.closed {
+		p.Close()
+		return mangos.ErrClosed
+	}
+	h.workq[conn] = true
+	go h.worker(conn)
+	return nil
+}
+
+func (h *connHandshaker) Close() error {
+	h.Lock()
+	h.closed = true
+	h.cv.Broadcast()
+	for conn := range h.workq {
+		conn.Close()
+	}
+	for len(h.doneq) != 0 {
+		item := h.doneq[0]
+		h.doneq = h.doneq[1:]
+		item.c.Close()
+	}
+	return nil
+}
+
+func (h *connHandshaker) worker(conn connHandshakerPipe) {
+	item := &connHandshakerItem{c: conn}
+	item.e = conn.handshake()
```

Comment:
If the `conn.handshake()` never returns due to a bad connection (as in #96), the `worker()` goroutine would leak.

Reply:
handshake should fail eventually.  Arguably this might be a case where the timeouts would be useful ... to detect a stuck TCP session.  That said, normally that shouldn't be an issue.  Typically the kernel will eventually detect that the TCP peer is dead.

Having said that, I'd be willing to set some kind of reasonable timeout on the handshake in particular.  That's probably a different case.  (Probably the timeout should be a small number of seconds -- under normal circumstances the handshake should complete within a few round trip times.)

Reply:
I'd like to handle setting the timeout for this (to close the potential leak of the file and associated go routine) separately.  That "leak" (I'd argue its not really a leak, as the peer could actually wake up and complete the operation some long time - days even! - later) is a far less bad situation than we have where a the bad peer would actually gum up everything immediately, which this PR fixes.